google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java

```diff
@@ -1767,17 +1717,15 @@ public BlobInfo internalCreateFrom(Path path, BlobInfo info, Opts<ObjectTargetOp
 
   @Override
   public BlobInfo internalDirectUpload(BlobInfo info, Opts<ObjectTargetOpt> opts, ByteBuffer buf) {
-
     BlobInfo.Builder builder =
-        info.toBuilder()
-            .setMd5(
-                BaseEncoding.base64().encode(Hashing.md5().hashBytes(buf.duplicate()).asBytes()))
-            .setCrc32c(
-                BaseEncoding.base64()
-                    .encode(Ints.toByteArray(Hashing.crc32c().hashBytes(buf.duplicate()).asInt())));
+        opts.blobInfoMapper().apply(info.toBuilder().clearMd5().clearCrc32c());
+    @Nullable Crc32cLengthKnown hash = opts.getHasher().hash(buf.duplicate());
```

Comment:
Does this create a copy of the data? Can you get a hash without doing that?

Reply:
`ByteBuffer#duplicate` creates a new buffer but shares the underlying bytes https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#duplicate--

This allows us to not have to manually rewind the buffer after it's been consumed by hashing.